common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/session/AbstractGatewaySessionHandler.java

```diff
@@ -368,24 +365,26 @@ void processOnDisconnect(MqttPublishMessage msg, String deviceName) {
 
     protected void onDeviceTelemetryJson(int msgId, ByteBuf payload) throws AdaptorException {
         JsonElement json = JsonMqttAdaptor.validateJsonPayload(sessionId, payload);
-        if (json.isJsonObject()) {
-            JsonObject jsonObj = json.getAsJsonObject();
-            for (Map.Entry<String, JsonElement> deviceEntry : jsonObj.entrySet()) {
-                String deviceName = deviceEntry.getKey();
+        if (!json.isJsonObject()) {
+            throw new JsonSyntaxException(CAN_T_PARSE_VALUE + json);
+        }
+
+        for (Map.Entry<String, JsonElement> deviceEntry : json.getAsJsonObject().entrySet()) {
+            if (!deviceEntry.getValue().isJsonArray()) {
+                log.warn("{}[{}]", CAN_T_PARSE_VALUE, json);
+                continue;
+            }
+
+            String deviceName = deviceEntry.getKey();
+            T deviceCtx = devices.get(deviceName);
```

Comment:
we should use a single method 
checkDeviceConnected(deviceName)

That returns future (immediate or real future)

Then check if the future is completed and successful - process now or add a callback

Having a custom device check (accessing internal map) looks like we have duplicated the logic